File: #152_Maximum_Product_Subarray/Solution.java
public class Solution {
    public int maxProduct(int[] nums) {
		if (nums.length == 0) return 0;
		int localMin = nums[0];
		int localMax = nums[0];
		int globalMax = nums[0];
		for (int i = 1; i < nums.length; i++) {
			int prevMin = localMin;
			localMin = Math.min(nums[i], Math.min(localMin * nums[i], localMax * nums[i]));
			localMax = Math.max(nums[i], Math.max(prevMin * nums[i], localMax * nums[i]));
			globalMax = (globalMax > localMax) ? globalMax : localMax;
		}
		return globalMax;
    }

    public static void main(String[] args) {
    	int[] nums = {1, -2, 3, -4, 5, -6};
    	Solution sol = new Solution();
    	System.out.println(sol.maxProduct(nums));
    }
}

File: #152_Maximum_Product_Subarray/Solution.py
'''
O(n) time Solution
'''

class Solution:
	# @param {integer[]} nums
	# @return {integer}
	def maxProduct(self, nums):
		if not nums:
			return 0
		maxVal = [0 for x in range(len(nums))]
		minVal = [0 for x in range(len(nums))]
		maxVal[0] = nums[0]
		minVal[0] = nums[0]
		re = maxVal[0]
		for x in range(1, len(nums)):
			maxVal[x] = max(maxVal[x - 1] * nums[x], minVal[x - 1] * nums[x], nums[x])
			minVal[x] = min(maxVal[x - 1] * nums[x], minVal[x - 1] * nums[x], nums[x])
			re = max(re, maxVal[x])
		return re

sol = Solution()
print sol.maxProduct([-2])